projects/ti_to_ti/pre_tokenize_parallel_example.py

```diff
@@ -0,0 +1,317 @@
+        f"{assistant_prompt.format(output='')}"
+    )
+    
+    if len(input_images)==0 and len(output_images)==0 :
+        return {
+            'text': text,
+            'prompt': prompt,
+            'input_image': None,
+            'image': None,
+        }
+    elif len(input_images)==0:
+        return {
+            'text': text,
+            'prompt': prompt,
+            'input_image': None,
+            'image': output_images,
+        }
+    else:
+        return {
+            'text': text,
+            'prompt': prompt,
+            'input_image': input_images,
+            'image': input_images + output_images,
+        }
+    
+def format_sample(raw_sample: dict[str, Any]) -> dict[str, Any]:
+    if raw_sample['source'] == "LLaVA-150K":
+        return format_sample_llava(raw_sample)
+    else:
+        return format_sample_cham(raw_sample)
+
+
+def preprocess(tokenizer, processor, formatted_sample: dict[str, Any]):
+    return_dict = {}
+    raw_text = ''
+    if isinstance(formatted_sample['text'], list):
+        raw_text = tokenizer.eos_token.join(formatted_sample['text'])
+    elif isinstance(formatted_sample['text'], str):
+        raw_text = formatted_sample['text'] + tokenizer.eos_token
+    else:
+        raise NotImplementedError
+        
+    if formatted_sample['image'] is None:
+        text_dict = processor(raw_text, return_tensors='pt').to(dtype = torch.bfloat16)
+        return_dict['input_ids'] = text_dict['input_ids'].squeeze()
+        formatted_prompt = formatted_sample['prompt']
+        prompt_dict = processor(formatted_prompt, return_tensors='pt').to(dtype = torch.bfloat16)
+        labels = return_dict['input_ids'].clone()
+        
+        # print(len(prompt_dict['input_ids'][0]))
+        labels[: len(prompt_dict['input_ids'][0] - 1)] = IGNORE_INDEX
+        return_dict['labels'] = labels
+        return_dict['pixel_values'] = None
+        return return_dict, len(prompt_dict['input_ids'][0] - 1), False
+    else: 
+        text_dict = processor(raw_text, formatted_sample['image'], return_tensors='pt').to(dtype = torch.bfloat16)
+        
+        return_dict['input_ids'] = text_dict['input_ids'].squeeze()
+
+        formatted_prompt = formatted_sample['prompt']
+        if formatted_sample['input_image'] is None:
+            prompt_dict = processor(formatted_prompt, return_tensors='pt').to(dtype = torch.bfloat16)
+        else:
+            prompt_dict = processor(formatted_prompt, formatted_sample['input_image'], return_tensors='pt').to(dtype = torch.bfloat16)
+        
+        labels = return_dict['input_ids'].clone()
+        # mask non-assistant input
+        labels[: len(prompt_dict['input_ids'][0]) - 1] = IGNORE_INDEX
+        return_dict['labels'] = labels
+
+        return_dict['pixel_values'] = text_dict['pixel_values']
+        
+        # print(f"Input id shape: {return_dict['input_ids'].shape}")
+        # print(f"Pixel value shape: {return_dict['pixel_values'].shape}")
```

Comment:
delete

Reply:
fixed in https://github.com/PKU-Alignment/align-anything/pull/41/commits/068a8bbbaf4a27c191a98f19007a7f354d922d7b